engine/src/main/java/org/pentaho/di/trans/streaming/common/AtomicCountSemaphore.java

```diff
@@ -0,0 +1,79 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2020 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+package org.pentaho.di.trans.streaming.common;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Atomic Count Semaphore
+ * <br>
+ * This class implements just a subset of the functionality of {@link Semaphore} but provides faster performance. Also,
+ * {@link Semaphore#release(int)} in testing was found not to always be thread safe.
+ *
+ * Note that this implementation has a very narrow usage scope with the following restrictions:
+ * 1. Acquire may only be called by one thread, otherwise more permits can be acquired than permitCount
```

Comment:
Is it the thread that *creates* the semaphore that is allowed to call `.acquire`?  If so, should lock that down with a precondition.

Reply:
removed class